### ISSUE TITLE
Separate AssertionHelper project or assembly

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#1212]
Issue #1211 brings up the question. Maintaining a separate syntax using AssertionHelper is a bit of a burden and we have not kept it up to date. Perhaps it's time to drop it.

I'm creating this issue in case there is a silent fan base for AssertionHelper, so that folks may speak up in its favor. If we keep it, we should update it, but let's decide whether we want to keep it first,

A suboption would be to remove it from NUnit and publish it as a separate addon.

**Comment by rprouse** [comment]
I didn't even know we had an AssertionHelper :fearful:

**Comment by CharliePoole** [comment]
I take that as a vote for removal. :-)

**Comment by oznetmaster** [comment]
I never knew we had one either :(

**Comment by oznetmaster** [comment]
What does it do, or what is it supposed to do?

**Comment by CharliePoole** [comment]
Your test class inherits from it, giving you a syntax for assertions that doesn't require a static class name. See, for example, https://github.com/nunit/nunit-csharp-samples/blob/master/syntax/AssertSyntaxTests.cs and scan for "Expect"

As I commented in #1211, it has not been kept up to date with the latest features of the Assert.That syntax. Ideally, if we keep it, I think it should be spun off into a separate project and maintained by people who actually use it.

**Comment by rprouse** [comment]
> As I commented in #1211, it has not been kept up to date with the latest features of the Assert.That syntax. Ideally, if we keep it, I think it should be spun off into a separate project and maintained by people who actually use it.

I agree. How many ways to we want to support writing asserts? Keeping Assume in sync with Assert is hard enough :smile:

**Comment by appel1** [comment]
In many cases you can use "using static nunit.framework.Assert" instead I think. So I don't mind if it was removed.

**Comment by OsirisTerje** [comment]
Agree with the above,  didn't know it existed either.  
Remove or move it out just for safekeeping.

**Comment by OsirisTerje** [comment]
Could we extract the helper in a nuget package - just create one manually for now since it is a kind of probe, and then see how many would download it ?  
If it gets  lets say less than 1000 over a 6 month period, then we drop it.

**Comment by MSK61** [comment]
I use it frequently in my projects, but I'm strongly in favor of phasing it out of NUnit(by either complete removal or moving to subsidiary package). I can easily refactor my projects to do without it; it's just a couple of find and replace for Expect(and the compiler will take care of the rest flagging errors where appropriate).

**Comment by CharliePoole** [comment]
Looks like a feature. :-)

We could do this in an adhoc way, but I'd like to use it as a pilot for something that was in the original design for NUnit 3. There was this idea that you could layer additional syntaxes on top of the framework assembly. This one is pretty uncomplicated and we might use it to come up with a design for alternate syntaxes added on to NUnit. That would mean it would become a separate project and download, as Terje suggests.

In and of itself, this is a low-priority feature, but we might push it forward in order to break out the design needed to layer on top of the framework. Should be pretty simple to do.

**Comment by CharliePoole** [comment]
@MSK61 Alternatively, are you interested in helping to maintain it if we move it to a separate package?

**Comment by CharliePoole** [comment]
PR #1250 has now isolated the use of AssertionHelper within our own tests. It is no longer used at all except in AssertionHelperTests. This enables us to easily separate it or remove it. 

I have marked this issue 'status-design' since we need to actually decide what we want to do before proceeding. Here's my take...

I don't think we should eliminate it because at least some people use it. The fact that it's a small number doesn't so much matter to me. We went to 3.0 without deciding on this as a breaking change so I would argue it is too late to do so now. NUnit has a long tradition of continuing to support users way beyond what most commercial projects do. I'd like to keep the tradition.

Further, I think that deciding how to handle AssertionHelper is a useful thing for us to do. The design of NUnit 3.0 originally called for alternative syntactic layers on top of the core of the framework. We set that aside for the initial release but I'd like to get back to it. AssertionHelper is relatively trivial - two files plus one test file - and we can concentrate on the desired architecture for syntactic front ends as a part of doing this.

[In case you have not seen the design docs for a while, syntactic adapters for the framework were intended to serve two purposes: alternative syntaxes for using NUnit and alternative language front ends. Existing 3rd party apps in each category are shouldly and fsunit. We planned to provide a standard way for creating syntax adapters on top of the framework, use it ourselves and offer it to 3rd party developers as a way of encouraging NUnit use. As stated above, this was too big a challenge to get into 3.0, so we have not implemented it.]

Back to AssertionHelper... here are some options...

* Keep it as a part of NUnit
    * Separate namespace (I don't like this because it breaks compatibility)
    * Separate assembly
* Make a new repo
    * Source code only - folks can include it in their own projects
    * Distribute Binaries
    * Do both (similar to what fsUnit does - this is the option I like best)

If our direction includes the distribution of binaries, we have to decide which builds to support. It could be all our existing builds, of course, but I'd be OK with limiting it to .NET 4.5 and Portable. We could add more if their were a demand. Users never tell us if we do too much, but they usually speak up if we do too little. :-)

I would favor moving to use of separate configurations rather than separate projects for each build type. That would limit the number of projects we have to define and we could use conditional compilation as needed. AssertionHelper includes a few features that have to be excluded in the Portable build.

Please add your comments here. When it looks like we have a consensus, I'll move ahead.

**Comment by ChrisMaddock** [comment]
This seems like a great way to 'deprecate' functionality, by moving it in to a separate repo and add-in packages. It's then clearly a separate maintenance concern, and can be updated just when new features are specifically asked for. (Which based on this topic, may be rarely!) It would also be significantly easier to break out than the other framework builds that are on the list, which would potentially create maintenance problems. 

To revitalise the discussion...

I think a separate repo, assembly and nuget package sounds ideal. It would require a new project reference for users ('breaking change') - but seems acceptable, only similar to what's just been required for .NET 2.0 with NUnit.Linq.

> Separate namespace (I don't like this because it breaks compatibility)

Agreed the current namespace should be maintained, which is a shame if it's going to be an add-in, but the compatibility issues would outweigh the 'ugliness' in my opinion!

>  It could be all our existing builds, of course, but I'd be OK with limiting it to .NET 4.5 and Portable.

For 'legacy' functionality, would it perhaps make sense to include/only do the older targets? 

> I would favor moving to use of separate configurations rather than separate projects for each build type. That would limit the number of projects we have to define and we could use conditional compilation as needed. AssertionHelper includes a few features that have to be excluded in the Portable build.

I assume by this you mean, a single VS project, and instead target it to different runtimes at compile time, i.e. via Cake? If so, I agree.

**Comment by rprouse** [comment]
Won't this still be tightly coupled to the framework because of its need for the various assertions under the covers? If that is the case, then a separate assembly would link to a specific version of the framework and would have to be released in sync. If that is the case, then isn't it easier to leave it where it is?

**Comment by ChrisMaddock** [comment]
I presumed there would need to be something equivalently to nunit.engine.api at framework level? I wasn't sure if that was the "syntactic adapter" idea @CharliePoole was referring to.

**Comment by rprouse** [comment]
I'm sure we can figure it out with some sort of indirection like the engine
uses, I just wonder if it is worth the effort :-)

**Comment by CharliePoole** [comment]
Well, it's low priority, so not worth a ton of effort. :-) Something to consider when we decide how to do it.

@ChrisMaddock I've been using priorities slightly differently for issues that are part of Epics. I mark those low that may end up being optional as far as the Epic goes.

@rprouse I think some sort of indirection is possible but that would make it dependent on a much more ambitious refactoring of the framework, similar to what the xunit guys have done. I've thought about that but not created an issue beyond just separating assertions. Maybe it's 4.0 material. :-)

@ChrisMaddock Yes... I was suggesting we have configs like "Portable - Debug", "Desktop - Release", etc. Or maybe we could use the new netstandard definitions instead. 

For a low-pri task like this, my first choice would be to let it sit for a while. My second choice, if somebody is dying to give it a shot, would be to make a separate assembly within the existing solution. But really, I'd rather see it wait until after the epic is complete.

@rprouse I was going to post my proposed order of doing the sub-issues on the Epic but was waiting for your comments on it first. Should I just go ahead and post?

**Comment by rprouse** [comment]
@CharliePoole go ahead and post. We've talked about it enough that we are both on the same page.

**Comment by CharliePoole** [comment]
Looking back at this after a long absence, here are my present-day thoughts...

1. We should __really__ get rid of this. It gives an alternative way to do Asserts, but doesn't replace Assume or Warn, so it's a partial implementation, which we do not want to complete.

2. My concerns about changing the namespace may have been unwarranted. Since you can only use this by inheriting from AssertionHelper, you really only need to change the declaration line for your test class.

3. We need to decide if we want to move this out incrementally or in one big change. Is it better for users to suffer multiple changes or just one change, even if slightly bigger?

**Comment by ChrisMaddock** [comment]
One big change sounds good.

Were you still considering a separate package, at least for a few releases to see what usage is like?

**Comment by CharliePoole** [comment]
Added this back to the backlog since we are only marking it obsolete for now.

**Comment by ChrisMaddock** [comment]
If we do go forward and remove this - the samples repo's will also need updating. They use AssertionHelper.

**Comment by CharliePoole** [comment]
We can do that in advance. I added issues for C#, C++ and VB. The F# samples don't use it.

**Comment by fluffynuts** [comment]
Just leaving my 2c here:

I, for one, am a big fan of the Expect syntax provided by AssertionHelper, especially when one has to deal with Javascript projects alongside .NET ones and the syntax remains familiar between the two.

It seems as if interest in AssertionHelper is the realm of the esoteric -- so I volunteer to be part of maintaining AssertionHelper if it is to be released as a separate repo. An idea would be to also release same-versioned nuget packages (ie nunit x.y.z has a corresponding nunit.assertionhelper x.y.z which depends specifically on nunit x.y.z) in parallel with nunit releases. Again, I volunteer to be a part of this.

An idea also might be to provide Expect as a static which can be statically imported, so that test fixtures don't have to inherit from AssertionHelper -- which makes some other scenarios easier to deal with. In fact, this sounds (to me) like a relatively clean way to make the break. Thoughts?

Since discovering AssertionHelper about 6 or 8 months ago, I've actually been migrating tests towards that syntax, whenever I come across tests using the static Assert syntax, so I'm actually quite invested in keeping this alive, if possible.

**Comment by CharliePoole** [comment]
@fluffynuts It's great if you want to become the maintainer of AssertionHelper and I'd be glad to help you do it. It's a little harder to make you __part__ of maintaining it if the NUnit Core Team is removing it from NUnit.

@ChrisMaddock @jnm2 I think your 👍 's are a little ambiguous. Are you all for dropping `AssertionHelper` (which I think you both voted for) and having somebody else take it over? Or are you for @fluffynuts request that we keep it? (provided I understood correctly) 😄

**Comment by fluffynuts** [comment]
@CharliePoole well, if the NUnit Core Team is set on removing it, perhaps the easiest solution is that I fork? And at some point I could implement the static Expect which would enable `using static` and mean that fixtures wouldn't have to inherit from AssertionHelper any more. 

If I were to do this, I would be interested in is some kind of notification from the NUnit Core Team of an upcoming release, such that I could try to release a same-versioned nuget package depending on the new release as soon as possible after the NUnit release. I'd be happy to do the work involved though, as long as everyone on the NUnit Core Team is ok with that. Obviously then I'd be fielding any requests against the Expect syntax, but I'd also have the huge advantage of starting off with a rich, working codebase, not trying to do an Expect-a-like.

Thoughts?

**Comment by CharliePoole** [comment]
Let's give the thumbs up guys a chance to respond. I've stated the direction that I __thought__ we had decided to go, but it's always possible - especially working remotely - to have completely opposite understandings out of the same discussion.

Regarding `using static` I'm not clear how that would give you any of the advantages of AssertionHelper __except__ for the verb `Expect`. How would you deal with constraints?

**Comment by fluffynuts** [comment]
To be honest, I haven't given much thought to the idea beyond "it would be
cool if I could get the Expect syntax without inheritance". I'll give it a
closer look tomorrow (:

**Comment by ChrisMaddock** [comment]
What I wanted to see was the AssertionHelper separated from the framework - we have a lot to maintain here and a big API, and, in my opinion, it's the right limb to chop off. **If** there is interest (which it seems there is from at least one user!) - a separate NuGet package which depends on the framework package seems an ideal solution to me.

As I understood, one of the motivations for deprecating, rather than straight removal which we also discussed was to get a proper idea of usage - and @fluffynuts has come forward with exactly the sort of proposition we talked about - to maintain AssertionHelper as a separate package. Hence my 👍!

I'd be happy for it to stay in the NUnit **organization**, separate from the framework. This would depend on any future maintainer being willing, of course - I'm not volunteering to look after it! Should for whatever reason maintenance drop off, I could see it going the same way as the SL and CF builds - unmaintained, but archived, and ready for anyone who wished to pick up.

**Comment by CharliePoole** [comment]
@ChrisMaddock Thanks for clarifying. All I remembered was that I had made it a separate project but backed out that change because somebody (maybe not you) wanted us to either keep it or kill it completely.

**Comment by jnm2** [comment]
My thoughts are the same as @ChrisMaddock's.
Also it does change the picture if we don't know of anyone who is motivated to invest in it, and then someone does come along. Even if I think we should spend no more time on it, I'm never going to vote to block a motivated party from working on something like this.

**Comment by rprouse** [comment]
I am also fine with someone else taking over and making it a separate package that depends on the main framework. I am also happy to split it out into it's own repo under the main org if @fluffynuts wants to maintain and release it. At the very least, if it is a separate NuGet package, we can track downloads and with that usage.

**Comment by CharliePoole** [comment]
@jnm2 Remember that this is open source. Even if we just erased it, it still exists in the repo and anybody is free to fork it and release it themselves. We don't really have to decide if the software will live, just whether we want to work on it or take responsibility for it.

That said, I think it's cleaner if we make it in to it's own project first. Then somebody can fork it as they like OR they can maintain it as part of our stable of projects. I had it completely split off but in the code review it was decided not to do that so I backed out the change. However, it's pretty easy to do it again if we want to.

I think it's best if we take steps we won't have to reverse later. So, for example, if we make it a separate assembly, then a separate package and later stop maintaining it, that's a kind of natural order. If we jump ahead too fast and then have to roll it back it's a waste of effort.

**Comment by CharliePoole** [comment]
So what are we doing with this?

**Comment by fluffynuts** [comment]
@CharliePoole I'm ok with whatever "sits best" with the NUnit Core Team (:

It looks like there's a general leaning towards splitting out into a separate project with its own package -- cool.

What I'd like to achieve, if possible, is that that project releases packages in lock-step with the main NUnit project (as mentioned above: {new project} x.y.z should depend specificially on NUnit x.y.z -- to avoid confusion about which packages play well together and also so a lazy person like me can just `install-package {new project}` and get NUnit brought in with that.
I think it would be a win if this was kept as an "official" NUnit thing. I'm not sure how you guys would like to proceed on that front though. I'm also still willing to just fork and pull out AssertionHelper and release on my own, but that's not my preference, to be honest.
Also, if a new project were spawned (either by NUnit Core or myself), I'd like the name to at least be acceptable to the NUnit Core -- I don't want to fork and call it something easy and obvious like "NUnit.AssertionHelper" or (better, imo) "NUnit.Expectations" (because that describes what it does) and have the official NUnit bleak about me piggy-backing off of the official brand (:

Basically, I'm happy with whatever you guys choose to do and I'm still willing to maintain whatever animal emerges.

**Comment by jnm2** [comment]
> I think it would be a win if this was kept as an "official" NUnit thing. I'm not sure how you guys would like to proceed on that front though. I'm also still willing to just fork and pull out AssertionHelper and release on my own, but that's not my preference, to be honest.

@CharliePoole Btw this is what I meant by not blocking. I'd like this to be free to happen since @fluffynuts is motivated to invest in it.

**Comment by rprouse** [comment]
I am happy to keep it as an official NUnit project that will be maintained and released as long as the community continues to use it and maintain it. It also looks like we have general agreement from the @nunit/core-team. @CharliePoole do you want to re-instate the previous repo that you created for this, or would you prefer someone else do it?

**Comment by CharliePoole** [comment]
@jnm2 To be clear - in case I wasn't earlier - I am for this idea. I'm raising questions that have to do with the overall running of the project.

Several people in the discussion are members of the Core Team, which will be making major decisions for the project in the future. It has only been a matter of months since they decided __not__ to accept my proposal to separate AssertionHelper into it's own repo.

Aside from the personal annoyance I feel at having discarded that work, I don't think the Core Team will work well if it's basic policies change every month.

The fact that someone is saying he likes Assertion helper is not new. We always knew it had its followers. 

The fact that someone has volunteered to work on it is new but shouldn't drive policy in my opinion. A board like the Core Team should not be reactive but active. If this was a good idea, we should have been working to implement it. That includes actively seeking a maintainer.

I realize this governance business is a new thing, so glitches are inevitable. But I didn't detect any awareness of this being a complete reversal from just a short while back.

And of course @jnm2 has only just joined the Core Team.

@fluffynuts Sorry too make your proposal an example. I'm in favor of a separate repo and - as you have probably figured out - already did the work to create one a little while back.

**Comment by rprouse** [comment]
I thought that our original plan was to not move in 3.7 because it was a breaking change for users and we wanted to give them time to update and allow us to gauge the reaction. I thought removing it without first deprecating it for a release or two wasn't a good idea.

I thought that we would then decide if we wanted to move it to a new project and find a maintainer or remove it completely. We have someone that wants to maintain it, so I am fine with allowing @fluffynuts to take it on. The end result is that it will be removed from the main project as we were planning and the separate project will either continue if it is supported or die a natural death if it isn't.

**Comment by ChrisMaddock** [comment]
This was also my understanding.

**Comment by CharliePoole** [comment]
@rprouse @ChrisMaddock I don't want to beat a dead horse here, but if that was your understanding, then we should not have merged a change that marked AssertionHelper obsolete, with the notation that it was going to be removed in a future release. I submitted that based on my understanding of what you guys were saying. If we go ahead with the separate repo, then we should remove the ObsoleteAttribute.

Do we have some difference of opinion about what Obsolete means? Serious question. To me it means that the feature is __definitely__ going away. I would think that's how users take it too. I would not use it to mean we wish it would go away or it might go away.

@rprouse Reinstate? AFAIK, there's nothing left to reinstate. I'll look in GitHub, but I believe all that stuff got squashed out. Locally, I deleted the code because it was not going to be needed - we were removing the feature. This was a change I really wanted to see, but I was voted down, pretty much unanimously. I apologize for my rancor about it, which I didn't actually know was there until now.

If I can find any remnants of the change, I will put them back up in a PR. If not, I prefer to work with @fluffynuts to help him set it up with a standard build, etc.

**Comment by CharliePoole** [comment]
I'm going to reserve any further "policy" discussion for either email or the governance repo.

**Comment by jnm2** [comment]
If AssertionHelper is moving to a new NuGet package and a new assembly, it's actually a good thing we obsoleted it from the NUnit framework assembly. We just need to update the obsoletion message now telling people where to go for the correct assembly.

**Comment by CharliePoole** [comment]
@jnm2 Obsolete is for stuff that's going away entirely. If the new class - wherever it is - is going to have the same name, then we should not use Obsolete. Where you find the class is a matter for some other warning, probably in the release where the change happens.

On the other hand, @fluffynuts has suggested some new names to replace AssertionHelper and I agree that a new name is a good idea. In that case, "AssertionHelper" __is__ going away and Obsolete is sensible. The message could say "Use SomeOtherClass, found in some.other.assembly"

Of course, this only matters if we do a hotfix release. If we wait for 3.8, it may already be gone.

**Comment by jnm2** [comment]
I obsolete things that move between assemblies because now you have access to two identical APIs available to IntelliSense at the same time and the obsoletion message would be helpful, saying don't use `AssertionHelpers, NUnit.Framework`, instead use `AssertionHelpers, NUnit.Expectations`. From an API  compatibility viewpoint, changing the assembly is exactly the same break and obsoletion process as changing the name. Of course we're free to interpret things however we choose. We don't have to follow that. I was just recognizing that as a good thing which I've seen before.

**Comment by CharliePoole** [comment]
@jnm2 That's only possible if the two exist at the same time. I was thinking they would not. I suppose we could leave a dummy AssertionHelper in the framework and mark it obsolete. Is that what you are thinking?

In any case, we can't do that right now, because the separate version of AssertionHelper does not yet exist. Right now (3.7.1) is what I was talking about. What would you use during the period where the new assertion helper (or whatever name) doesn't yet exist?

**Comment by CharliePoole** [comment]
@fluffynuts IF you are going to be doing this, I suggest you propose a plan for the sequence of things. Our 3.8 release will likely not be for another 3 months. We may have one or more hotfix releases - Rob has scheduled 3.7.1 for a few days from now.

Even without enhancements and working quickly, I can't imagine you will be at the point of creating a release of the new assertion helper sooner than a few weeks from now - maybe a month.

Can you propose a sequence of steps?

**Comment by jnm2** [comment]
> What would you use during the period where the new assertion helper (or whatever name) doesn't yet exist?

How about we either keep the message that AssertionHelpers is going away from NUnit.Framework, which is true, or we go beyond that and say that we have tentative (or firm) plans for it to be moved to NUnit.Expectations.

**Comment by fluffynuts** [comment]
@CharliePoole well, one way of doing things (not necessarily the _right_ way, but one way (: ) would be to fork the NUnit repo, under the NUnit account, grant me maintainer/developer access and I clone, wantonly ripping out everything that isn't AssertionHelper or something which directly supports it, making the hosting project for AssertionHelper depend on an NUnit nuget package. I reckon that wouldn't take too long.
The stickier part comes in with ensuring that what I've done is acceptable to the NUnit team and then using your build and release process. If I were to quickly hack it and release on my nuget account, that wouldn't take long -- but I don't think that's the right way to go, personally. I'd like control of this new repo and the packages produced from it to exist outside of myself. I don't mind having joint control, but I certainly don't want to have it appear on Nuget as one of _my_ packages, making it such that if I can't maintain it any more, it can't just be handed over to someone else. The same rationale applies to the repo.
So, a possible sequence of events might be:

1.  Fork NUnit under the NUnit account
2. Grant me developer / maintainer access
3. I hack and slash to get the source code to a place where it's really just an unmodified AssertionHelper with only what it needs to survive, depending on an NUnit nuget package
4. In the meantime, the Obsolete message is updated to refer to installing a different package to get the required functionality. This message should include the final package name -- I'm happy with NUnit.Expectations
5. Hopefully I can get a release out soonish and there's only a minor hiccup for end-users. If there are "a few days", I should be able to complete the code side, assuming a lightning fast fork at your side. Then it's up to code review and ensuring my build and release align with yours

I agree that Obsolete should be for features which are definitely going to die, but this is (imo) an acceptable use of the attribute, if for no other reason than that there isn't some existing attribute which would offer the same functionality: a compiler warning with instructions on how to fix it.

Thoughts?

**Comment by CharliePoole** [comment]
@fluffynuts I located a local backup copy of the nunit-assertion-helper repository. I can rename it nunit-expectations before uploading it if everyone likes that name. I have to look at the code to see what's there but it looks like it has a proper build etc. This is pure chance due to my not having purged it from one of my machines yet.

@nunit/framework-team @nunit/core-team 
Unless I hear otherwise, I'll make the repo nunit-expectations and the assembly nunit.expectations.dll. I'll upload what I have to master of the new repo - it's unchanged from the framework AFAIK. Further changes can be dealt with in code review.

@fluffynuts 
There really is not a bunch of hacking and slashing to do to remove AssertionHelper. PR #1250 isolated it completely in a single file - with tests in another file. It is not referred to by any other part of NUnit. The new repo is what we decided not to do a few months back, but I kept all the changes that isolated the feature.

We have two possibilities here:
1. The 3.7.1 hotfix release will go out before this is ready.
2. This will be ready to release at the same time as the hotfix.

I don't think we should remove AssertionHelper in the hotfix in either case. That's not what a hotfix is for. But we should probably change the message on the ObsoleteAttribute and it will be different depending on whether the situation is 1 or 2. I think we can assume 2 if you want to do a release with only the name changes. If you want to add some features, then it's 1. However, bear in mind, you can do releases as frequently as you want. The only glitch is that it's going to be easier for everyone in this case if the release numbers match those for the framework you are referencing.

**Comment by fluffynuts** [comment]
@CharliePoole epic win! (:

I think that the first release of the package should literally just be to cleave it from NUnit, ie change nothing other than the assembly name and where the code lives (:

I would prefer, as mentioned before, to do package releases in lock-step with NUnit main, matching version numbers -- it will be simpler for downstream users to deal with. It's for this reason that I re-release all of my own PeanutButter.* packages whenever I update any of them: I've had users ask me long ago what versions will work with each other and it's just way simpler to match version numbers.

For interim feature releases, -rc[X] could be used in the version (basically using the [pre-release mechanism](https://docs.microsoft.com/en-us/nuget/create-packages/prerelease-packages) -- in particular, if I'm working on something new, I'd want to install the bleeding-edge into my own projects: nothing play-tests an idea better than eating your own dogfood.

**Comment by ChrisMaddock** [comment]
Sounds good to me. 

Unless namespace/class names are changing, it wouldn't be possible to use the new AssertionHelper package, while the old AssertionHelper still exists in the framework, would it? 

That suggest to me that 3.7.1 just changes the obsolete message to mention the new package, and 3.8 removes the old assertionhelper, and is the first release of the new package.

@CharliePoole - I wonder if our different ideas of 'obsolete' is where the confusion arose from. Either way - glad you've managed to recover the work.

**Comment by CharliePoole** [comment]
In the new repo, I put it into the NUnit.Framework.Extensions namespace. If that's OK with everyone, then any user could switch to the new version at any time by use of an alias.

I created the repo and I'm working on changing names of assemblies and projects. Will add package ref to 3.7 and then request a review. @fluffynuts can continue from there.

**Comment by jnm2** [comment]
@CharliePoole Who is the current assignee for this issue?

**Comment by CharliePoole** [comment]
Nobody. I was taking the tack that issues are local to repos, with this one about removing assertion helper being a framework issue. I created the new repo, nunit-expectations. If it's wanted, I can do some more work on it and we can file some issues there.

The practical side of this is that you can't really do an integration test with the new component until the framework is changed and you can't change the framework till the new component is ready. 😄 

My recommendation was for this to be in the 3.8 release.

**Comment by CharliePoole** [comment]
I should add... I'm assuming this will be a contributed project, which normally means somebody comes in as the project lead rather than one of our committers. **However** if one of our newer (than me or Rob) team members wanted to take this on as a first shot of being a project lead, I think it would be a good experience, maybe leading to bigger things. 😺 🥇 😈

**Comment by jnm2** [comment]
@CharliePoole I'm just wondering what remaining work is keeping _this_ issue from being closed and if we know who will do it.

**Comment by CharliePoole** [comment]
@jnm2 This issue is about removing the files from the framework and documenting how people should deal with the change. It's trivial but somebody has to decide when to do it. We probably don't want it to be up-for-grabs, since it requires coordination with the other project that replaces the feature. 

I should probably get the other project building so somebody can take it over in conjunction with this issue.

**Comment by CharliePoole** [comment]
@rprouse Would you want this done right before release of 3.8 or is any time during the period OK?

**Comment by rprouse** [comment]
Wow, 60 comments. I think we are spending more time talking about things than coding 😈 

If @fluffynuts wants to take over maintenance, then how about we give him the chance to fork and prepare a pull request. I will create a repo for him to work in and we can see how it goes, just say the word.

As for removing AssertionHelper from this project, we just deprecated it in 3.7. How about we wait a release to give people time to update their code? It will also give @fluffynuts a chance to get something out. If he moves quickly, then we can always re-evaluate for the 3.8 release since users will have something to switch to.

**Comment by CharliePoole** [comment]
@rprouse I agree about the talking.

Maybe there are so many comments that the essence is getting lost. Why do we need a fork? We already did all the work and merged it... minus dropping two files. The separate repo already exists. 

I believe it actually waits for you to say __when__ you want it to happen, there's hardly any work left to do and the risk is that it gets done too soon and then stagnates if you are not ready to have it merged.

I assumed you would want it gone in 3.8 and was asking if you wanted to wait to do it in the release branch or would you want it done now, so it would disappear for anyone using dev builds.

**Comment by rprouse** [comment]
@CharliePoole I clearly missed something while I was catching up. I thought that the you had lost your work on the separate repo. Where is it now? I don't see it  under the nunit org or your account. If that repo exists and we can help @fluffynuts create a release, then I think we can proceed with removing AssertionHelper from this repo and make the separate AssertionHelper release just before 3.8.

If we don't have a separate release of AssertionHelper available, I would prefer to wait another release to give people more time to update.

I just don't think we should be doing the work of maintaining AssertionHelper. If the community wants it and steps up, then it allows us to remove it from the framework earlier because there will be an easy alternative. If the community doesn't step up, then we give people a bit of time and it is gone in 3.9.

**Comment by CharliePoole** [comment]
See https://github.com/nunit/nunit/issues/1212#issuecomment-311964899

Summary: I found a backup of the local repository I created and uploaded it to https://github.com/nunit/nunit-expectations. We merged PR #1250 back February, which eliminated all of our own use of AssertionHelper and isolated it to two files.

I think I should take a look to refresh my memory and create  as needed on nunit-expectations... i.e. does it build, etc.

**Comment by fluffynuts** [comment]
Hi all

This has been open for a while. I know I offered to maintain AssertionHelper, but I may have found a way that makes everyone happy -- the people who want to move it out as well as people (like me) who currently use the syntax.

I put up for review [https://github.com/fluffynuts/NUnit.StaticExpect](https://github.com/fluffynuts/NUnit.StaticExpect) which is up on [nuget.org](https://www.nuget.org/packages/NUnit.StaticExpect/). I've used this to replace all usages of AssertionHelper in one project -- though my usage may not be as extensive as others, so I welcome feedback.

Basically, it's just a wrapper around `Assert.That` calls, but it did provide a drop-in replacement for AssertionHelper, at least, for me. And it achieves my goal of not requiring all test fixtures to inherit from another class.

It also provides a stop-gap for me as I finish up a first release of [NExpect](https://github.com/fluffynuts/NExpect) which aims to provide a more Chai-like syntax (eg `Expect(value).To.Equal(otherValue)`) with a Jasmine-like experience for consumer extension. It dove-tails with NUnit, without depending directly on it, throwing AssertionExceptions if the type is found at run-time, otherwise throwing it's own exceptions.

So far, it's shaping up, but I want to include a few more features before a 1.0. If any Expect users are interested, perhaps either watch this project or let me know so I can keep you updated.

**Comment by CharliePoole** [comment]
@fluffynuts That looks like a good alternative. Do you know if other IDEs and compilers support the syntax?


As for our own NUnit support, what I see happening is that I'll make sure the new project is building after my vacation, then leave the schedule of switching over to it up to @rprouse and the framework team.

**Comment by fluffynuts** [comment]
@CharliePoole `using static` is part of the C#6 spec [https://msdn.microsoft.com/en-us/magazine/dn802602.aspx](https://msdn.microsoft.com/en-us/magazine/dn802602.aspx) so it _should_ be supported by anything which supports C#6 syntax (: 

Whilst there might compilers that I don't know about which don't support this, certainly any modern Visual Studio will support this out of the box. Older Visual Studio versions (and I'm assuming other processes using msbuild) can get support through the `Microsoft.Net.Compilers` nuget package (and then they also get C#7 support -- which is a good thing too (: ).
Mono supports C#6: [http://www.mono-project.com/docs/about-mono/languages/csharp/](http://www.mono-project.com/docs/about-mono/languages/csharp/) and I'm making the bold assumption that dotnet core, being the new kid on the block, supports a spec from 2014 (:

So whilst I _think_ this should be a viable alternative for everyone, I'm also quite open to being told that I'm wrong (:

**Comment by CharliePoole** [comment]
So it's at least possible for the framework team to remove this feature entirely, provided that they don't mind the loss of backward compatibility for what is probably a small number of users.

Personally, I would continue to maintain the separate project, but not enhance it in the future.

**Comment by fluffynuts** [comment]
I agree -- whilst it doesn't impact me, NUnit has a very wide reach and I'm sure there's someone it would impact.

**Comment by fluffynuts** [comment]
After discussion with Bob Sammers, who has made valuable contribution to NUnit.StaticExpect, I'd just like to remind anyone subscribed to, or stumbling across this thread that [NUnit.StaticExpect](https://github.com/fluffynuts/NUnit.StaticExpect) (nuget: [https://nuget.org/packages/NUnit.StaticExpect](https://nuget.org/packages/NUnit.StaticExpect)) provides a fairly effortless drop-in replacement for AssertionHelper. The GitHub page is also a good place to raise issues and discussion, if required.

I didn't want to be a troll on this forum, but Bob pointed out that perhaps people didn't realise the completeness of the solution and how easy it is to switch over, so I'm just putting that out there again. Under the hood, it's still making "proper" NUnit calls, so this really just provides the syntactic sugar AssertionHelper users have been used to whilst allowing them to prepare for the eventual removal of the class.

**Comment by ChrisMaddock** [comment]
Great to see NUnit.StaticExpect getting some attention @fluffynuts - this is a great solution to the problem!

-----

@nunit/framework-team - I think we're probably close to being able to wrap this issue up. Can I propose the following, based on where we are now?

1. AssertionHelper remains obsolete and unmaintained for the rest of the 3.x series.
1. AssertionHelper is removed in 4.0
1. We update documentation and the ObsoleteAttribute to recommend NUnit.StaticExpect as a drop-in replacement.
1. NUnit.StaticExpect lives on as a third-party library, which now provides an _alternate_ syntax, as opposed to a 'legacy' syntax. We previously talked about doing this within the NUnit org - I don't see any reason to bring this into the organisation any more. @fluffynuts - if you disagree at all, I'd personally have no objection to moving this into the org - but I think what's key is that your remain maintainer as an independent project. 🙂

**Comment by fluffynuts** [comment]
@ChrisMaddock I don't have any problem with bringing this into the NUnit organisation at all -- indeed, I think it might be a good idea because it will look like a more "official" alternative instead of some code cranked out by an arbitrary code-monkey on the interwebs (: Not sure how this is achieved though, so I welcome information (:

I also *intend* to remain as maintainer -- but it might be a good idea, if there's enough interest in the project, to provide full access to someone else (perhaps @BobSammers, who has contributed already), for those typical "run over by a bus" moments.

The same should be done for package management on nuget.org, for the same reasons.

**Comment by ChrisMaddock** [comment]
@fluffynuts Ok! 🙂

Let's discuss the future of NUnit.StaticExpect separately, in [governance](https://github.com/nunit/governance) - and keep this issue for deciding what to do about the current NUnit.Framework.AssertionHelper - so we can get this boxed off. I'm out of time for now, but will write up an issue tonight to start the discussion.

**Comment by CharliePoole** [comment]
Keeping to the topic of this issue, how does the existence of `NUnit.StaticExpect` impact our decision about `AssertionHelper`? I had not heard of it before and it looks like an excellent solution. I was ready to drop `AssertionHelper` cold - even without a replacement - and still am. What about other @nunit/framework-team members? Does `NUnit.StaticExpect` as an alternative lead you to be ready to drop it?

As a minor matter, I would not call this a "drop-in" replacement since the user has to change the code. It's a small change, but still a change. I suspect it might only be a replacement for C#  users as well. In any case, I'm hoping it's enough of a replacement to cause us to accelerate dropping `AssertionHelper`.

**Comment by jnm2** [comment]
> Does NUnit.StaticExpect as an alternative lead you to be ready to drop it?

If our release notes, and what people will see conspicuously in the NuGet package manager, show a warning at the top of the description like, "Warning: the obsolete AssertionHelper has been removed. To continue using it, you must also reference the NUnit.StaticExpect package." — then personally I'm equally happy doing it next minor release or continuing to wait.

**Comment by CharliePoole** [comment]
@jnm2 The notice would have to be more like "reference the NUnit.StaticExpect package" and

* stop inheriting from AssertionHelper
* add a using statement
* possibly some code fixes (@fluffynuts is there anything else?)

I'd still drop it myself. IMO StaticExpect's use of static methods instead of inheritance is a significant improvement.

**Comment by fluffynuts** [comment]
@CharliePoole I can't think of anything to add, offhand, though most of what you mention should be in the README.md already (and if there's anything which isn't, it *should* be added; indeed, this README *should*, imo, provide all the information a person needs to switch -- because it would be silly and annoying if it didn't!). 

So if you have the time, I'd appreciate a review and commentary on [https://github.com/fluffynuts/NUnit.StaticExpect/blob/master/README.md](https://github.com/fluffynuts/NUnit.StaticExpect/blob/master/README.md) such that it incorporates, in an easy-to-follow manner, the steps required to move from AssertionHelper to NUnit.StaticExpect. If that README.md becomes clear and definitive enough, it might be good enough to make the deprecation notice something like "AssertionHelper has been deprecated in favor of NUnit.StaticExpect. Read https://github.com/fluffynuts/NUnit.StaticExpect/blob/master/README.md to find out more". 
Thoughts?

**Comment by CharliePoole** [comment]
@fluffynuts The only thing I'd want to see added would be a few more examples, e.g.:

* Use of constraints (replacing `Is` and `Has` syntax)
* Use with other than Assert/Expect: e.g. Assume, Warn.

I just wanted to be sure it's a complete replaceme (or give notice if it isn't)

What's the deal with non-C# code? Can you use it in VB for example?

**Comment by CharliePoole** [comment]
@fluffynuts To be clear, I'd dump `AssertionHelper` without a replacement but I've been in the minority. But we recommend a replacement, I'd like to make it completely clear what it replaces and what (if anything) it doesn't. I like the idea that your readme should explain everything.

**Comment by fluffynuts** [comment]
@CharliePoole I'm totally on-board with the idea that a readme that the user is directed towards should be explanatory and clear. I'm also aware that I may miss something which you (or anyone else on the NUnit team) might deem important -- hence the call for review. To that end, and to answer some of the issues you raise:

- I'll certainly add some more examples
- Assume / Warn appear to not be part of the `AssertionHelper` syntax, so they are currently not implemented in `NUnit.StaticExpect`. Currently, we have two testing methods for proving parity:   reflection-based testing which shows that (in all cases but 2) that the methods & properties that would be found on `AssertionHelper` are mimicked via static methods & properties on `Expectations` (with the two mentioned having special-case behavioral tests since they have diverged from the exact signature offered by `Assert.That`) as well as the full `AssertionHelper` test fixture running against `Expecations` instead of inheriting from `AssertionHelper`. The primary aim of the project is to provide a minimal-pain replacement for `AssertionHelper`, so that's pretty-much where the current state is. Of course, `Assume` and `Warn` are still available to the user via regular calls.
- non-C# code: this is a good question. It's all .net, so, sure, you can use it from, say, VB; but I think your question is more around static imports in VB.NET -- a question I can't answer right now (and which my google-fu has failed to answer so far), but which I will find a definitive answer for by simply creating a project and trying. I should be able to report back here later today on this.

And to re-iterate, for clarity: I appreciate review, commentary and suggestions on the README because it really should be a friendly, informative landing-point for anyone considering the switch to `NUnit.StaticExpect`. All suggestions welcome.

**Comment by fluffynuts** [comment]
@CharliePoole to my glee:
`Imports NUnit.StaticExpect.Expectations`
provides the same functionality in VB.NET as
`using static NUnit.StaticExpectations.Expectations`
does in C#. Just tested this now in a small VB.NET test library.

**Comment by fluffynuts** [comment]
(and, as a side-note, I now understand why `AssertionHelper` copied off the methods on (for example), `Is` - because I can't, in VB.Net, do 
`Expect(1, Is.EqualTo(1))`,
but I can (through the static import) do:
`Expect(1, EqualTo(1))`

So I just need to flesh out the documentation more. I can give that a look tonight (:

**Comment by CharliePoole** [comment]
I was only thinking you might want to let people know they can use `Assume` or `Warn` along with the static constraint-building properties. 

WRT `Is` in VB - that's why we have `Iz`. 😄

**Comment by BobSammers** [comment]
As @fluffynuts mentioned, `StaticExpect` tests against the `AH` functionality via reflection to ensure complete coverage of `AH`. Currently, 2 of these tests fail (actually are special-cased out) because the functionality of `AH` has diverged from the `Is` and `Does` members that it shadows (specifically `InRange` and `Exactly`, respectively).

I'd be interested in how long `AH` is likely to remain in NUnit. If it's likely to be for a couple  more releases (which makes sense if sign-posts to StaticExpect can be put in to its obsoletion message), would you be interested in taking a pull request for changes to `AssertionHelper.InRange()` and `AssertionHelper.Exactly()` to close the divergence (I'm assuming `AH` was left behind accidentally)? This would give `AH` and `SE` exactly equivalent in functionality before `AH` is deprecated (and help with the `SE` test suite!).

**Comment by CharliePoole** [comment]
@BobSammers We haven't been adding features to `AssertionHelper` for a while, so it doesn't surprise me that it has fallen behind some new features. Feel free to create an issue for any divergence you want to fix... I put it that way, because it probably wouldn't get much priority otherwise.

@nunit/framework-team There hasn't been much recent comment here, so it's hard to respond to @fluffynuts and @BobSammers in a meaningful way. It seems like we "sort of" decided to get rid of `AssertionHelper` in the past and I __think__ we were planning to direct users to the separate assembly I created. But we suffer a bit from the lack of a clear decision, it seems.

I believe we have had two fundamentally different directions under discusssion up to now:

1. Stop support for `AssertionHelper` entirely.
2. Move it to a separate package, which we may later deprecate and remove.

All the other options we discussed seem to me to be just details in one case or the other. Most of them have to do with how fast we implement one strategy or the other.

The availablility of `StaticExpect` seems to make alternative (2) a bit silly. There is no reason for us to compete with `StaticExpect`, especially since it looks like a better solution than inheritance. I suppose that this could open up a third option, if we want to continue supporting the API

3. Incorporate `StaticExpect` into NUnit, replacing `AssertionHelper`. In fact, NUnit had a static class named `ConstraintFactory` until quite recently, which could have been used in much the same way as `StaticExpect` if any of us had thought of it. 😞 

@rprouse This is looking to me like one of those strategic decisions that needs to be made one way or the other by the project lead. After that, we can debate the details.

**Comment by fluffynuts** [comment]
Whichever way things go, I've updated the README.md tonight, largely borrowing from the AssertionHelper documentation.
As a side-note I've also update the solution to use a single, multi-target project, so it's a little neater. And included generated XML documentation in the package.

**Comment by ChrisMaddock** [comment]
> @nunit/framework-team There hasn't been much recent comment here, so it's hard to respond to @fluffynuts and @BobSammers in a meaningful way. 

Sorry Charlie - but it's only been a little over 24 hours! Afraid I can't always respond to things that fast. 😞 

----

Previously, I was keen on just removing AssertionHelper in a minor release. What changed my mind was the issues we saw with removing TestFixtureSetUpAttribute etc in 3.8. To summarise - in 3.8 we removed some attributes which had been deprecated since 3.0 - assuming this would be a trivial find/replace fix for any users still relying on these. What we found out is that certain third party libraries were still relying on these attributes - meaning users of these 3rd party libraries were unable to update NUnit, until the library was patched. Of course - that's not impossible, but it meant that what we assumed would be a trivial fix, was not at all for a number of users. The same situation may well apply here.

AssertionHelper is now entirely segregated from the rest of the codebase thanks to @CharliePoole's work - hence my preference now for just leaving it tucked away in the corner to rot until 4.0. We can recommend users migrate to NUnit.StaticExpect anyway. 

---

Regarding the points you laid out Charlie, I agree (2) would now be silly. On (3) - I'm personally keen on thinning out the core API we're providing, and think NUnit.StaticExpect would be a great third-party alternative syntax, rather than something that should be part of the core library. My overall preference would still be as I proposed [above](https://github.com/nunit/nunit/issues/1212#issuecomment-337187015) - but I think we'll need to agree to disagree there!

I'll also kick off this project contribution issue in governance tonight - sorry, ran out of time there yesterday.

**Comment by ChrisMaddock** [comment]
Separate issue for discussing bringing NUnit.StaticExpect into the NUnit organisation: https://github.com/nunit/governance/issues/33

**Comment by CharliePoole** [comment]
@ChrisMaddock Actually, you're the one who responded already! That said, we seem to have left this undecided for something like nine months and we've been silent since July, until @fluffynuts just pushed on it again.

I like your proposal with one reservation. I think we should modify the `Obsolete` message to recommend use of `StaticExpect` now, rather than waiting.

FWIW, my desire to remove this sooner was predicated on the notion that I had to be around as part of the team so as to add my replacement project. I just wanted to get a pending project off my plate. If we can drop a separate `AssertionHelper` package as an option, I'm happy to have the feature hang around till 4.0.

**Comment by ChrisMaddock** [comment]
> I like your proposal with one reservation. I think we should modify the Obsolete message to recommend use of StaticExpect now, rather than waiting.

I agree - these were meant as three separate points - not an ordered list. 😄

**Comment by mikkelbu** [comment]
I think that https://github.com/nunit/nunit/issues/1212#issuecomment-337695330 very precisely captures my thoughts (and is expressed much better than I can do), especially regarding the "thinning out the core API".
